### ISSUE TITLE
Unexpected warnings when special functions are explicitly compared with boolean values

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#29855]
```
        [Fact]
        public void NotNullWhenFalse_EqualsTrue()
        {
            CSharpCompilation c = CreateCompilation(new[] { @"
using System.Runtime.CompilerServices;
public class C
{
    public void Main(string? s)
    {
        if (MyIsNullOrEmpty(s) == true)
        {
            s.ToString(); // warn
        }
        else
        {
            s.ToString(); // ok
        }

        s.ToString(); // warn 2
    }
    public static bool MyIsNullOrEmpty([NotNullWhenFalse] string? s) => throw null;
}
", NotNullWhenFalseAttributeDefinition, NonNullTypesTrue, NonNullTypesAttributesDefinition });

            // PROTOTYPE(NullableReferenceTypes): there should only be two diagnostics
            c.VerifyDiagnostics(
                // (9,13): warning CS8602: Possible dereference of a null reference.
                //             s.ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13),
                // (13,13): warning CS8602: Possible dereference of a null reference.
                //             s.ToString(); // ok
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(13, 13),
                // (16,9): warning CS8602: Possible dereference of a null reference.
                //         s.ToString(); // warn 2
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(16, 9)
                );
        }

        [Fact]
        public void NotNullWhenFalse_EqualsFalse()
        {
            CSharpCompilation c = CreateCompilation(new[] { @"
using System.Runtime.CompilerServices;
public class C
{
    public void Main(string? s)
    {
        if (false == MyIsNullOrEmpty(s))
        {
            s.ToString(); // ok
        }
        else
        {
            s.ToString(); // warn
        }

        s.ToString(); // warn 2
    }
    public static bool MyIsNullOrEmpty([NotNullWhenFalse] string? s) => throw null;
}
", NotNullWhenFalseAttributeDefinition, NonNullTypesTrue, NonNullTypesAttributesDefinition });

            // PROTOTYPE(NullableReferenceTypes): there should only be two diagnostics
            c.VerifyDiagnostics(
                // (9,13): warning CS8602: Possible dereference of a null reference.
                //             s.ToString(); // ok
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13),
                // (13,13): warning CS8602: Possible dereference of a null reference.
                //             s.ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(13, 13),
                // (16,9): warning CS8602: Possible dereference of a null reference.
                //         s.ToString(); // warn 2
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(16, 9)
                );
        }

        [Fact]
        public void NotNullWhenFalse_NotEqualsFalse()
        {
            CSharpCompilation c = CreateCompilation(new[] { @"
using System.Runtime.CompilerServices;
public class C
{
    public void Main(string? s)
    {
        if (false != MyIsNullOrEmpty(s))
        {
            s.ToString(); // warn
        }
        else
        {
            s.ToString(); // ok
        }

        s.ToString(); // warn 2
    }
    public static bool MyIsNullOrEmpty([NotNullWhenFalse] string? s) => throw null;
}
", NotNullWhenFalseAttributeDefinition, NonNullTypesTrue, NonNullTypesAttributesDefinition });

            // PROTOTYPE(NullableReferenceTypes): there should only be two diagnostics
            c.VerifyDiagnostics(
                // (9,13): warning CS8602: Possible dereference of a null reference.
                //             s.ToString(); // warn
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(9, 13),
                // (13,13): warning CS8602: Possible dereference of a null reference.
                //             s.ToString(); // ok
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(13, 13),
                // (16,9): warning CS8602: Possible dereference of a null reference.
                //         s.ToString(); // warn 2
                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(16, 9)
                );
        }
```

**Comment by gafter** [comment]
In flow analysis, there are no split states ("when true" and "when false") as a result of the binary operators, with the exception of `&&` and `||`.  This is by design.  One can imagine relaxing that for the nullable analysis, especially given that this analysis is not intended to be precise.  However, we do expect the analysis to be specified, and it isn't clear that there is sufficient benefit to extending the analysis to the boolean operators `==` and `!=`.  Extending it could be very complicated.  For example, what if you compare the result of two special bool-returning functions using `==`?
